powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/internal/IdempotencyHandler.java

```diff
@@ -0,0 +1,159 @@
+import software.amazon.lambda.powertools.idempotency.exceptions.*;
+import software.amazon.lambda.powertools.idempotency.persistence.BasePersistenceStore;
+import software.amazon.lambda.powertools.idempotency.persistence.DataRecord;
+import software.amazon.lambda.powertools.utilities.JsonConfig;
+
+import java.time.Instant;
+
+import static software.amazon.lambda.powertools.idempotency.persistence.DataRecord.Status.EXPIRED;
+import static software.amazon.lambda.powertools.idempotency.persistence.DataRecord.Status.INPROGRESS;
+
+/**
+ * Internal class that will handle the Idempotency, and use the {@link software.amazon.lambda.powertools.idempotency.persistence.PersistenceStore}
+ * to store the result of previous calls.
+ */
+public class IdempotencyHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(IdempotencyHandler.class);
+    private static final int MAX_RETRIES = 2;
+
+    private final ProceedingJoinPoint pjp;
+    private final JsonNode data;
+    private final BasePersistenceStore persistenceStore;
+
+    public IdempotencyHandler(ProceedingJoinPoint pjp, String functionName, JsonNode payload) {
+        this.pjp = pjp;
+        this.data = payload;
+        persistenceStore = Idempotency.getInstance().getPersistenceStore();
+        persistenceStore.configure(Idempotency.getInstance().getConfig(), functionName);
+    }
+
+    /**
+     * Main entry point for handling idempotent execution of a function.
+     *
+     * @return function response
+     */
+    public Object handle() throws Throwable {
+        // IdempotencyInconsistentStateException can happen under rare but expected cases
+        // when persistent state changes in the small time between put & get requests.
+        // In most cases we can retry successfully on this exception.
+        for (int i = 0; true; i++) {
+            try {
+                return processIdempotency();
+            } catch (IdempotencyInconsistentStateException e) {
+                if (i == MAX_RETRIES) {
+                    throw e;
+                }
+            }
+        }
+    }
+
+    /**
+     * Process the function with idempotency
+     *
+     * @return function response
+     */
+    private Object processIdempotency() throws Throwable {
+        try {
+            // We call saveInProgress first as an optimization for the most common case where no idempotent record
+            // already exists. If it succeeds, there's no need to call getRecord.
+            persistenceStore.saveInProgress(data, Instant.now());
+        } catch (IdempotencyItemAlreadyExistsException iaee) {
+            DataRecord record = getIdempotencyRecord();
+            return handleForStatus(record);
+        } catch (IdempotencyKeyException ike) {
+            throw ike;
+        } catch (Exception e) {
+            throw new IdempotencyPersistenceLayerException("Failed to save in progress record to idempotency store", e);
```

Comment:
Since this is a catch all exception, I prefer adding a helpful text as well, something on the lines of,

> If you believe this is a bug, please open an issue here